controller/system_backup_controller.go

```diff
@@ -0,0 +1,1092 @@
+	}
+
+	if systemBackup.Status.OwnerID != c.controllerID {
+		systemBackup.Status.OwnerID = c.controllerID
+		systemBackup, err = c.ds.UpdateSystemBackupStatus(systemBackup)
+		if err != nil {
+			// we don't mind others coming first
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+
+		log := getLoggerForSystemBackup(c.logger, systemBackup.Name, "")
+		log.Infof("Picked up by SystemBackup Controller %v", c.controllerID)
+	}
+
+	log := getLoggerForSystemBackup(c.logger, systemBackup.Name, backupTargetClient.GetURL())
+
+	record := &systemBackupRecord{}
+	existingSystemBackup := systemBackup.DeepCopy()
+	defer c.handleStatusUpdate(record, systemBackup, existingSystemBackup, err, log)
+
+	if !systemBackup.DeletionTimestamp.IsZero() &&
+		systemBackup.Status.State != longhorn.SystemBackupStateDeleting {
+		c.updateSystemBackupRecord(record,
+			systemBackupRecordTypeNormal, longhorn.SystemBackupStateDeleting,
+			constant.EventReasonDeleting, SystemBackupMsgDeletingRemote,
+		)
+		return
+	}
+
+	// Sync the ready SystemBackups that are not in the current cluster.
+	if systemBackup.Labels != nil && systemBackup.Labels[types.GetVersionLabelKey()] != "" && systemBackup.Status.State == longhorn.SystemBackupStateNone {
+		c.updateSystemBackupRecord(record,
+			systemBackupRecordTypeNormal, longhorn.SystemBackupStateSyncing,
+			constant.EventReasonSyncing, SystemBackupMsgSyncingBackupTarget,
+		)
+		return nil
+	}
+
+	archievePath := filepath.Join(SystemBackupTempDir, systemBackup.Name+types.SystemBackupExtension)
+	tempDir := filepath.Join(SystemBackupTempDir, systemBackup.Name)
+
+	switch systemBackup.Status.State {
+	case longhorn.SystemBackupStateSyncing:
+		err = syncSystemBackupFromBackupTarget(systemBackup, backupTargetClient)
+		if err != nil {
+			c.updateSystemBackupRecord(record,
+				systemBackupRecordTypeError, longhorn.SystemBackupStateError,
+				longhorn.SystemBackupConditionReasonSync, SystemBackupErrSync,
+			)
+			return err
+		}
+
+		c.updateSystemBackupRecord(record,
+			systemBackupRecordTypeNormal, longhorn.SystemBackupStateReady,
+			constant.EventReasonSynced, SystemBackupMsgSyncedBackupTarget,
+		)
+
+	case longhorn.SystemBackupStateNone:
+		err = c.InitSystemBackup(systemBackup, log)
+		if err != nil {
+			return err
+		}
+
+		c.updateSystemBackupRecord(record,
+			systemBackupRecordTypeNormal, longhorn.SystemBackupStateGenerating,
+			constant.EventReasonStart, SystemBackupMsgStarting,
+		)
+
+	case longhorn.SystemBackupStateGenerating:
+		go c.GenerateSystemBackup(systemBackup, archievePath, tempDir)
+
+	case longhorn.SystemBackupStateUploading:
+		go c.UploadSystemBackup(systemBackup, archievePath, tempDir, backupTargetClient)
+
+	case longhorn.SystemBackupStateReady, longhorn.SystemBackupStateError:
+		cleanupLocalSystemBackupFiles(archievePath, tempDir, log)
+
+	case longhorn.SystemBackupStateDeleting:
+		err = c.cleanupRemoteSystemBackupFiles(systemBackup, backupTargetClient)
```

Comment:
If the system backup was failed with error state before, how this remote deletion will work for it? Should assume it work if the remote backup not found.

Reply:
Added verification to check if the backup exists in the backup target.